.30 AOT fails with  Module build failed: TypeError: Cannot read property 'kind' of undefined

### OS
> Windows 7, 8 or 10. Linux (which distribution). Mac OSX (Yosemite? El Capitan?)

```
@angular/cli: 1.0.0-beta.30
node: 6.9.5
os: win32 x64
@angular/cli: 1.0.0-beta.30
@angular/common: 2.4.6
@angular/compiler: 2.4.6
@angular/core: 2.4.6
@angular/forms: 2.4.6
@angular/http: 2.4.6
@angular/language-service: 2.4.6
@angular/platform-browser: 2.4.6
@angular/platform-browser-dynamic: 2.4.6
@angular/router: 3.4.5
@angular/compiler-cli: 2.4.6
@ngtools/webpack: 1.2.8
```


### Repro steps.
With latest .30 getting errors with AOT mode:


```


ERROR in ./src/store/effects/appdb.effects.ts
Module build failed: TypeError: Cannot read property 'kind' of undefined
    at C:\Temp\ia\studioweb\node_modules\@ngtools\webpack\src\loader.js:93:31
    at Array.map (native)
    at _addCtorParameters (C:\Temp\ia\studioweb\node_modules\@ngtools\webpack\src\loader.js:92:46)
    at C:\Temp\ia\studioweb\node_modules\@ngtools\webpack\src\loader.js:114:17
    at Array.forEach (native)
    at _removeDecorators (C:\Temp\ia\studioweb\node_modules\@ngtools\webpack\src\loader.js:109:10)
    at C:\Temp\ia\studioweb\node_modules\@ngtools\webpack\src\loader.js:282:48
    at process._tickCallback (internal/process/next_tick.js:103:7)
 @ ./src/locale/app/app-module.ngfactory.ts 65:0-62
 @ ./src/main.ts
 @ multi ./src/main.ts

ERROR in ./src/store/effects/msdb.effects.ts
Module build failed: TypeError: Cannot read property 'kind' of undefined
    at C:\Temp\ia\studioweb\node_modules\@ngtools\webpack\src\loader.js:93:31
    at Array.map (native)
    at _addCtorParameters (C:\Temp\ia\studioweb\node_modules\@ngtools\webpack\src\loader.js:92:46)
    at C:\Temp\ia\studioweb\node_modules\@ngtools\webpack\src\loader.js:114:17
    at Array.forEach (native)
    at _removeDecorators (C:\Temp\ia\studioweb\node_modules\@ngtools\webpack\src\loader.js:109:10)
    at C:\Temp\ia\studioweb\node_modules\@ngtools\webpack\src\loader.js:282:48
    at process._tickCallback (internal/process/next_tick.js:103:7)
 @ ./src/locale/app/app-module.ngfactory.ts 66:0-61
 @ ./src/main.ts
 @ multi ./src/main.ts

ERROR in ./src/comps/sliderpanel/Sliderpanel.ts
Module build failed: TypeError: Cannot read property 'kind' of undefined
    at C:\Temp\ia\studioweb\node_modules\@ngtools\webpack\src\loader.js:93:31
    at Array.map (native)
    at _addCtorParameters (C:\Temp\ia\studioweb\node_modules\@ngtools\webpack\src\loader.js:92:46)
    at C:\Temp\ia\studioweb\node_modules\@ngtools\webpack\src\loader.js:114:17
    at Array.forEach (native)
    at _removeDecorators (C:\Temp\ia\studioweb\node_modules\@ngtools\webpack\src\loader.js:109:10)
    at C:\Temp\ia\studioweb\node_modules\@ngtools\webpack\src\loader.js:282:48
    at process._tickCallback (internal/process/next_tick.js:103:7)
 @ ./src/modules/shared.module.ts 2:0-63
 @ ./src/locale/app/app-module.ngfactory.ts
 @ ./src/main.ts
 @ multi ./src/main.ts

npm ERR! Windows_NT 10.0.14393
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "x_prod_aot"
npm ERR! node v6.9.5
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! studioweb@0.1.93 x_prod_aot: `ng build --target=production --base-href ./ --aot true`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the studioweb@0.1.93 x_prod_aot script 'ng build --target=production --base-href ./ --aot true'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the studioweb package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     ng build --target=production --base-href ./ --aot true
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs studioweb
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls studioweb
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\Temp\ia\studioweb\npm-debug.log

npm ERR! Windows_NT 10.0.14393
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "release_aot"
npm ERR! node v6.9.5
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! studioweb@0.1.92 release_aot: `rm -r -f ./dist && npm run x_bump && npm run x_prod_aot && npm run x_sw && npm run x_rsync`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the studioweb@0.1.92 release_aot script 'rm -r -f ./dist && npm run x_bump && npm run x_prod_aot && npm run x_sw && npm run x_rsync'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the studioweb package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     rm -r -f ./dist && npm run x_bump && npm run x_prod_aot && npm run x_sw && npm run x_rsync
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs studioweb
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls studioweb
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\Temp\ia\studioweb\npm-debug.log

root@DESKTOP-VEUHFOL /cygdrive/c/Temp/ia/studioweb
$


```

JIT works with both re-builds and lazy loading chunks.
but AOT is no go

anything I can try?

repo: 
git clone https://github.com/born2net/studioweb.git

regards

Sean

--- a/packages/@ngtools/webpack/src/loader.ts
+++ b/packages/@ngtools/webpack/src/loader.ts
@@ -61,35 +61,57 @@ function _angularImportsFromNode(node: ts.ImportDeclaration, sourceFile: ts.Sour
 function _ctorParameterFromTypeReference(paramNode: ts.ParameterDeclaration,
                                          angularImports: string[],
                                          refactor: TypeScriptFileRefactor) {
-  if (paramNode.type.kind == ts.SyntaxKind.TypeReference) {
-    const type = paramNode.type as ts.TypeReferenceNode;
-    const decorators = refactor.findAstNodes(paramNode, ts.SyntaxKind.Decorator) as ts.Decorator[];
-    const decoratorStr = decorators
-      .map(decorator => {
-        const fnName =
-          (refactor.findFirstAstNode(decorator, ts.SyntaxKind.CallExpression) as ts.CallExpression)
-          .expression.getText(refactor.sourceFile);
+  let typeName = 'undefined';
 
-        if (angularImports.indexOf(fnName) === -1) {
-          return null;
+  if (paramNode.type) {
+    switch (paramNode.type.kind) {
+      case ts.SyntaxKind.TypeReference:
+        const type = paramNode.type as ts.TypeReferenceNode;
+        if (type.typeName) {
+          typeName = type.typeName.getText(refactor.sourceFile);
         } else {
-          return fnName;
+          typeName = type.getText(refactor.sourceFile);
         }
-      })
-      .filter(x => !!x)
-      .map(name => `{ type: ${name} }`)
-      .join(', ');
-
-    if (type.typeName.kind == ts.SyntaxKind.Identifier) {
-      const typeName = type.typeName as ts.Identifier;
-      if (decorators.length > 0) {
-        return `{ type: ${typeName.text}, decorators: [${decoratorStr}] }`;
-      }
-      return `{ type: ${typeName.text} }`;
+        break;
+      case ts.SyntaxKind.AnyKeyword:
+        typeName = 'undefined';
+        break;
+      default:
+        typeName = 'null';
     }
   }
 
-  return 'null';
+  const decorators = refactor.findAstNodes(paramNode, ts.SyntaxKind.Decorator) as ts.Decorator[];
+  const decoratorStr = decorators
+    .map(decorator => {
+      const call =
+        refactor.findFirstAstNode(decorator, ts.SyntaxKind.CallExpression) as ts.CallExpression;
+
+      if (!call) {
+        return null;
+      }
+
+      const fnName = call.expression.getText(refactor.sourceFile);
+      const args = call.arguments.map(x => x.getText(refactor.sourceFile)).join(', ');
+      if (angularImports.indexOf(fnName) === -1) {
+        return null;
+      } else {
+        return [fnName, args];
+      }
+    })
+    .filter(x => !!x)
+    .map(([name, args]: string[]) => {
+      if (args) {
+        return `{ type: ${name}, args: [${args}] }`;
+      }
+      return `{ type: ${name} }`;
+    })
+    .join(', ');
+
+  if (decorators.length > 0) {
+    return `{ type: ${typeName}, decorators: [${decoratorStr}] }`;
+  }
+  return `{ type: ${typeName} }`;
 }
 
 
@@ -106,12 +128,7 @@ function _addCtorParameters(classNode: ts.ClassDeclaration,
   }
 
   const params = Array.from(ctor.parameters).map(paramNode => {
-    switch (paramNode.type.kind) {
-      case ts.SyntaxKind.TypeReference:
-        return _ctorParameterFromTypeReference(paramNode, angularImports, refactor);
-      default:
-        return 'null';
-    }
+    return _ctorParameterFromTypeReference(paramNode, angularImports, refactor);
   });
 
   const ctorParametersDecl = `static ctorParameters() { return [ ${params.join(', ')} ]; }`;

--- a/tests/e2e/assets/webpack/test-app/app/app.component.ts
+++ b/tests/e2e/assets/webpack/test-app/app/app.component.ts
@@ -1,4 +1,5 @@
 import {Component, ViewEncapsulation} from '@angular/core';
+import {MyInjectable} from './injectable';
 
 
 @Component({
@@ -7,4 +8,8 @@ import {Component, ViewEncapsulation} from '@angular/core';
   styleUrls: ['./app.component.scss'],
   encapsulation: ViewEncapsulation.None
 })
-export class AppComponent { }
+export class AppComponent {
+  constructor(public inj: MyInjectable) {
+    console.log(inj);
+  }
+}

--- a/tests/e2e/assets/webpack/test-app/app/injectable.ts
+++ b/tests/e2e/assets/webpack/test-app/app/injectable.ts
@@ -1,0 +1,8 @@
+import {Injectable, Inject, ViewContainerRef} from '@angular/core';
+import {DOCUMENT} from '@angular/platform-browser';
+
+
+@Injectable()
+export class MyInjectable {
+  constructor(public viewContainer: ViewContainerRef, @Inject(DOCUMENT) public doc) {}
+}

--- a/tests/e2e/tests/packages/webpack/test.ts
+++ b/tests/e2e/tests/packages/webpack/test.ts
@@ -1,7 +1,7 @@
 import {normalize} from 'path';
 import {createProjectFromAsset} from '../../../utils/assets';
 import {exec} from '../../../utils/process';
-import {expectFileSizeToBeUnder, replaceInFile} from '../../../utils/fs';
+import {expectFileSizeToBeUnder, replaceInFile, expectFileToMatch} from '../../../utils/fs';
 
 
 export default function(skipCleaning: () => void) {
@@ -9,13 +9,20 @@ export default function(skipCleaning: () => void) {
     .then(() => createProjectFromAsset('webpack/test-app'))
     .then(() => exec(normalize('node_modules/.bin/webpack'), '-p'))
     .then(() => expectFileSizeToBeUnder('dist/app.main.js', 420000))
-    .then(() => expectFileSizeToBeUnder('dist/0.app.main.js', 40000))
+    .then(() => expectFileSizeToBeUnder('dist/0.app.main.js', 10000))
     // test resource urls without ./
     .then(() => replaceInFile('app/app.component.ts',
       './app.component.html', 'app.component.html'))
     .then(() => replaceInFile('app/app.component.ts',
       './app.component.scss', 'app.component.scss'))
-    .then(() => exec(normalize('node_modules/.bin/webpack'), '-p'))
-    // test
+    // test the inclusion of metadata
+    // This build also test resource URLs without ./
+    .then(() => exec(normalize('node_modules/.bin/webpack')))
+    .then(() => expectFileToMatch('dist/app.main.js',
+      new RegExp('MyInjectable.ctorParameters = .*'
+               + 'type: .*ViewContainerRef.*'
+               + 'type: undefined, decorators.*Inject.*args: .*DOCUMENT.*'))
+    .then(() => expectFileToMatch('dist/app.main.js',
+      new RegExp('AppComponent.ctorParameters = .*MyInjectable'))
     .then(() => skipCleaning());
 }

